src/main/java/org/junitpioneer/jupiter/report/PioneerReportListener.java

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2015-2020 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter.report;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.xml.XMLConstants;
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.Marshaller;
+import javax.xml.transform.Source;
+import javax.xml.transform.stream.StreamSource;
+import javax.xml.validation.Schema;
+import javax.xml.validation.SchemaFactory;
+import javax.xml.validation.Validator;
+
+import org.junit.platform.engine.TestExecutionResult;
+import org.junit.platform.engine.reporting.ReportEntry;
+import org.junit.platform.launcher.TestExecutionListener;
+import org.junit.platform.launcher.TestIdentifier;
+import org.junit.platform.launcher.TestPlan;
+import org.xml.sax.SAXException;
+
+/**
+ * <p>This listener creates an additional report for pioneer extension and validates it against the "pioneerreport.xsd".
+ * This is done because the default junit test report can't / should not be extended.
+ * A dynamic test report which shall be used by all test tools is planed, but not yet realized, see
+ * <a href="https://github.com/ota4j-team/opentest4j/issues/9">issue #9 at opentest4j</a> for current progress.</p>
+ *
+ * <p>The report contains test containers and its test methods. For each object properties, based on published
+ * report entries are reported. This allows extension to report additional values in the report.
+ * for test cases the result of the test is (of course) reported.</p>
+ *
+ * @see <a href="https://junit.org/junit5/docs/current/api/org.junit.platform.launcher/org/junit/platform/launcher/TestPlan.html">JUnit Docs</a>
+ */
+public class PioneerReportListener implements TestExecutionListener {
+
+	TestPlan storedTestPlan;
+	PioneerReport pioneerReport;
+
+	ConcurrentHashMap<String, List<Property>> propertyCache = new ConcurrentHashMap<>();
+	ConcurrentHashMap<String, String> testStatusCache = new ConcurrentHashMap<>();
+
+	@Override
+	public void reportingEntryPublished(TestIdentifier testIdentifier, ReportEntry entry) {
+		String id = testIdentifier.getUniqueId();
+
```

Comment:
You are always going to add all of the properties, so why not do it like this:
```
String id = testIdentifier.getUniqueId();
propertyCache.putIfAbsent(id, new ArrayList<>());

entry.getKeyValuePairs().forEach((key, value) -> propertyCache.get(id).add(new Property(key, value)));
```

Reply:
Thank you. Another point that streams are strange (the stream skips - whyever -  the EntrySet. *shrug* Updated the code.